mathext/hypergeo.go

```diff
@@ -0,0 +1,231 @@
+	y2 := cephes.Hyp2f1(b, 1-c+b, 1-a+b, 1/z)
+	y2 *= math.Gamma(c) / math.Gamma(a) * math.Gamma(a-b) / math.Gamma(c-b) * math.Pow(-z, -b)
+	return y1 + y2
+}
+
+func eqn15_2_10(a, b, c, x float64) float64 {
+	t := a - math.Round(a)
+	var f2 float64
+	f1 := Hypergeo(t, b, c, x)
+	var f0 float64
+	mai := int(math.Round(math.Abs(a)))
+	if a < 0 {
+		f0 = Hypergeo(t-1, b, c, x)
+		t--
+		for n := 1; n < mai; n++ {
+			f2 = f1
+			f1 = f0
+			f0 = -(2*t-c-t*x+b*x)/(c-t)*f1 - t*(x-1)/(c-t)*f2
+			t--
+		}
+	} else {
+		f0 = Hypergeo(t+1, b, c, x)
+		t++
+		for n := 1; n < mai; n++ {
+			f2 = f1
+			f1 = f0
+			f0 = -((2*t-c-t*x+b*x)*f1 + (c-t)*f2) / (t * (x - 1))
+			t++
+		}
+	}
+	return f0
+}
+
+func eqn15_2_12(a, b, c, x float64) float64 {
+	t := c - math.Round(c)
+	var f2 float64
+	f1 := Hypergeo(a, b, t, x)
+	var f0 float64
+	mci := int(math.Round(math.Abs(c)))
+	if c < 0 {
+		f0 = Hypergeo(a, b, t-1, x)
+		t--
+		for n := 1; n < mci; n++ {
+			f2 = f1
+			f1 = f0
+			f0 = -t*(t-1-(2*t-a-b-1)*x)/t/(t-1)/(x-1)*f1 - (t-a)*(t-b)*x/t/(t-1)/(x-1)*f2
+			t--
+		}
+	} else {
+		f0 = Hypergeo(a, b, t+1, x)
+		t++
+		for n := 1; n < mci; n++ {
+			f2 = f1
+			f1 = f0
+			f0 = -(t*(t-1-(2*t-a-b-1)*x)*f1 + t*(t-1)*(x-1)*f2) / ((t - a) * (t - b) * x)
+			t++
+		}
+	}
+	return f0
+}
+
+func eqn15_3_3(a, b, c, z float64) float64 {
+	y := Hypergeo(c-a, c-b, c, z)
+	return math.Pow(1-z, c-a-b) * y
+}
+
+func eqn15_3_4(a, b, c, z float64) float64 {
+	y := Hypergeo(a, c-b, c, z/(z-1))
+	return math.Pow(1-z, -a) * y
+}
+
+func eqn15_4_1(m int, b, c, z float64) float64 {
+	// lzn is log(z^n / n!).
+	var lzn float64
+	var lznSign int = 1
+	// sum is the sum of the evaluated polynomial.
+	var sum float64 = 1
+	for n := 1; n <= m; n++ {
+		lzn += math.Log(math.Abs(z)) - math.Log(float64(n))
+		if z < 0 {
+			lznSign *= -1
+		}
+
+		var w float64
+		var sign int = 1
+		lphC, sn := lpochhammer(c, n)
+		if sn == 0 { // divide by zero
+			return math.NaN()
+		}
+		w -= lphC
+		sign *= sn
+
+		lphM, sn := lpochhammer(float64(-m), n)
+		if sn == 0 {
+			continue
+		}
+		w += lphM
+		sign *= sn
+
+		lphB, sn := lpochhammer(b, n)
+		if sn == 0 {
+			continue
+		}
+		w += lphB
+		sign *= sn
+
+		sum += float64(sign) * float64(lznSign) * math.Exp(w+lzn)
+	}
+	return sum
+}
+
+func lpochhammer(x float64, n int) (float64, int) {
+	var y float64
+	var sign int = 1
+	for k := range n {
+		xk := x + float64(k)
+		if xk == 0 {
+			return math.NaN(), 0
+		}
+
+		y += math.Log(math.Abs(xk))
+		if xk < 0 {
+			sign *= -1
+		}
+	}
+	return y, sign
+}
+
+// isCephesBug checks if the received arguments hit a Cephes bug.
+// If the bug is hit, isCephesBug returns the remedied value.
+//
+// This bug concerns Cephes' evaluation of mathext.Digamma on negative integers,
+// which is mathematically undefined.
+// For reference, scipy also found this workaround necessary in:
+// https://github.com/scipy/scipy/commit/1dfea85b34e8cbd868f9fc43345898084ba73aab
```

Comment:
```suggestion
// isCephesBug checks whether the received arguments hit a Cephes bug,
// occurring when |a| or |b| >> |c|.
// If the bug is hit, isCephesBug returns the remedied value.
//
// This bug concerns Cephes' evaluation of mathext.Digamma on negative integers,
// where it is mathematically undefined.
//
// For reference, scipy also found this workaround necessary in:
// https://github.com/scipy/scipy/commit/1dfea85b34e8cbd868f9fc43345898084ba73aab
```

Reply:
Thanks for the rewording, reads much better.
I have made a small modification to make the criteria where this bug manifests more accurate.